keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,544 @@
+readiness probes, which [use a different code path][probes2].
+
+This change would be [API-compatible][compatible]. If the field is not
+specified, we will default to the current behaviour, which uses the
+`terminationGracePeriodSeconds`.
+
+[compatible]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility
+[probes1]: https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L631-L655
+[probes2]: https://github.com/kubernetes/kubernetes/blob/e414d4e5c2ab85eada7eb8e80a363658e199968b/pkg/kubelet/prober/prober_manager.go#L55
+
+### Configuration example
+
+```yaml
+spec:
+  terminationGracePeriodSeconds: 3600
+  containers:
+  - name: test
+    image: ...
+
+    ports:
+    - name: liveness-port
+      containerPort: 8080
+      hostPort: 8080
+
+    livenessProbe:
+      httpGet:
+        path: /healthz
+        port: liveness-port
+      failureThreshold: 1
+      periodSeconds: 60
+      # New field #
+      gracePeriodSeconds: 60
+```
+
+### User Stories (Optional)
+
+N/A - bugfix
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+This should be a low-risk API change as it is backwards-compatible. If the
+field is not set, the current behaviour will be maintained.
+
+## Design Details
+
+The `livenessProbe.terminationGracePeriodSeconds` (or `startupProbe.*`) shall
+not be greater than the pod-level `terminationGracePeriodSeconds`, and we will
+add validation to check this.
+
+`initialDelaySeconds` is not relevant to this value; it indicates how long we
+should wait before probing the container, but does not have any relation to how
+long a probe should take or how long it will take to shut down the container
+upon a failed probe.
+
+`periodSeconds` and `timeoutSeconds` indicate how long we think a probe should
+take, but not how long it will take to shut down the container; hence, they
+also are not relevant to this value. It is suggested but not required that
+`livenessProbe.terminationGracePeriodSeconds` be less than these values.
+
+Finally, `failureThreshold` and `successThreshold` are not relevant to this
+value; they indicate the number of failures required to mark a probe as
+succeeded or failed, but not how long to give a container to gracefully shut
+down upon failure.
+
+### Test Plan
+
+This change will be unit tested for the API changes and
+backwards-compatibility.
+
+If added to the Probe API, this will also need to be tested in the readiness
+and startup probe cases.
+
+E2E integration tests will verify the correct behaviour for the current broken
+use case, where a `terminationGracePeriodSeconds` is set to a
+large value, a probe fails, and with `livenessProbe.gracePeriodSeconds` set,
+the container terminates quickly.
+
+### Graduation Criteria
+
+Because this is a bugfix with a compatible API change, similar to
```

Comment:
I think this requires a feature gate, and should not be treated different than any other change.

Its a new feature (decouple timeout), whereas the referenced example was a bug in existing feature.

The new field requires careful rollout because of n-2 kubelet skew requirements.

The API can only be supported when we know all kubelets have the logic, so it needs to be off-by default for at least 2 releases until all kubelets are up to same level as API server.

Reply:
Yeah, needs a feature gate, it's an API change.